### PR TITLE
GH-111848: Set the IP when de-optimizing

### PR DESCRIPTION
**Pull request by markshannon**
[2% faster for tier 2](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20231112-3.13.0a1%2B-4d1e295-PYTHON_UOPS)
Vs tier 1, now 5% slower. Main is [7% slower](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20231114-3.13.0a1%2B-a519b87-PYTHON_UOPS)

* Issue: gh-111848

### CODE REVIEW

File: Python/optimizer_analysis.c [hunk 1]
```diff
@@ -42,12 +36,16 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
...
-            if (_PyOpcode_opcode_metadata[opcode].flags & (HAS_ERROR_FLAG | HAS_DEOPT_FLAG) || opcode == _PUSH_FRAME) {
-                need_ip = true;
-            }
             if (_PyOpcode_opcode_metadata[opcode].flags & HAS_ESCAPES_FLAG) {
                 maybe_invalid = true;
+                if (last_set_ip >= 0) {
+                    buffer[last_set_ip].opcode = _SET_IP;
+                }
+            }
+            if ((_PyOpcode_opcode_metadata[opcode].flags & HAS_ERROR_FLAG) || opcode == _PUSH_FRAME) {
```
Comment (iritkatriel) [inline Python/optimizer_analysis.c]: we generate `OPCODE_HAS_ESCAPES` for this.

File: Python/optimizer.c [hunk 2]
```diff
@@ -569,7 +565,7 @@ translate_bytecode_to_trace(
             {
                 if (instr + 2 - oparg == initial_instr && code == initial_code) {
                     RESERVE(1, 0);
-                    ADD_TO_TRACE(_JUMP_TO_TOP, 0, 0);
+                    ADD_TO_TRACE(_JUMP_TO_TOP, 0, 0, 0);
```
Comment (iritkatriel) [inline Python/optimizer.c:568]: target is 0 because `_JUMP_TO_TOP` never deopts?

File: Python/ceval.c [hunk 3]
```diff
@@ -1077,6 +1078,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 // Jump here from _EXIT_TRACE
 exit_trace:
     _PyFrame_SetStackPointer(frame, stack_pointer);
+    frame->instr_ptr = next_uop[-1].target + _PyCode_CODE((PyCodeObject *)frame->f_executable);
```
Comment (iritkatriel) [inline Python/ceval.c:1081]: Maybe assert that target is not 0?

File: Python/optimizer.c (hunk 2 above)
Comment (markshannon) [reply Python/optimizer.c:568]: Yes.

File: Python/ceval.c (hunk 3 above)
Comment (markshannon) [reply Python/ceval.c:1081]: Sometimes exit the trace immediately upon making a call, if it appears to be recursive.
So, target can be zero.

**Review by iritkatriel — APPROVED** [review]

File: Python/optimizer.c [hunk 4]
```diff
@@ -458,23 +459,12 @@ translate_bytecode_to_trace(
     trace[trace_length].opcode = (OPCODE); \
     trace[trace_length].oparg = (OPARG); \
     trace[trace_length].operand = (OPERAND); \
+    trace[trace_length].target = (TARGET); \
     trace_length++;
 
 #define INSTR_IP(INSTR, CODE) \
     ((uint32_t)((INSTR) - ((_Py_CODEUNIT *)(CODE)->co_code_adaptive)))
 
-#define ADD_TO_STUB(INDEX, OPCODE, OPARG, OPERAND) \
```
Comment (gvanrossum) [inline Python/optimizer.c:466]: Ah, no more stubs. Do you expect stubs to eventually make a comeback? If not, there are a few mentions of 'stub' that can be removed.

Notably the `stub` arg to `RESERVE()` is now always 0. We could remove it.

**Pull request merged**